library/core/src/num/mod.rs

```diff
@@ -1032,29 +1028,52 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     // of multi-byte sequences
     let src = src.as_bytes();
 
-    let (is_positive, digits) = match src[0] {
-        b'+' | b'-' if src[1..].is_empty() => {
-            return Err(PIE { kind: InvalidDigit });
+    let (first, mut digits) = (*src.get(0).ok_or_else(|| PIE { kind: Empty })?, &src[1..]);
+
+    let (is_positive, mut result) = match first {
+        b'+' => {
+            let first = digits.get(0).ok_or_else(|| PIE { kind: InvalidDigit })?;
+            digits = &digits[1..];
+            (
+                true,
+                T::from_u32(
+                    (*first as char).to_digit(radix).ok_or_else(|| PIE { kind: InvalidDigit })?,
+                ),
+            )
         }
-        b'+' => (true, &src[1..]),
-        b'-' if is_signed_ty => (false, &src[1..]),
-        _ => (true, src),
+        b'-' if is_signed_ty => {
+            let first = digits.get(0).ok_or_else(|| PIE { kind: InvalidDigit })?;
+            digits = &digits[1..];
+            (
+                false,
+                T::from_u32(0)
+                    - T::from_u32(
+                        (*first as char)
+                            .to_digit(radix)
+                            .ok_or_else(|| PIE { kind: InvalidDigit })?,
+                    ),
+            )
+        }
+        val => (
+            true,
+            T::from_u32((val as char).to_digit(radix).ok_or(PIE { kind: InvalidDigit })?),
+        ),
     };
 
-    let mut result = T::from_u32(0);
-
-    if can_not_overflow::<T>(radix, is_signed_ty, digits) {
+    if mem::size_of::<T>() > 2 {
         // If the len of the str is short compared to the range of the type
         // we are parsing into, then we can be certain that an overflow will not occur.
         // This bound is when `radix.pow(digits.len()) - 1 <= T::MAX` but the condition
-        // above is a faster (conservative) approximation of this.
+        // in `safe_width` is a faster (conservative) approximation of this.
         //
         // Consider radix 16 as it has the highest information density per digit and will thus overflow the earliest:
```

Comment:
suggestion: this comment is useful information, but doesn't seem like it belongs here, since the computation it's talking about here isn't here.  Maybe put it in on/in `safe_width` instead?

Reply:
Said otherwise, this code will be correct as long as `safe_width` is correct, so the details of which approach -- faster or tighter -- doesn't really matter here.